src/modules/ekf2/EKF2.cpp

```diff
@@ -2625,9 +2626,81 @@ int EKF2::task_spawn(int argc, char *argv[])
+		int32_t acc2_prio = 0;
+		int32_t acc3_prio = 0;
+		int32_t mag0_prio = 0;
+		int32_t mag1_prio = 0;
+		int32_t mag2_prio = 0;
+		int32_t mag3_prio = 0;
+
+		param_get(param_find("CAL_ACC0_PRIO"), &acc0_prio);
+		param_get(param_find("CAL_ACC1_PRIO"), &acc1_prio);
+		param_get(param_find("CAL_ACC2_PRIO"), &acc2_prio);
+		param_get(param_find("CAL_ACC3_PRIO"), &acc3_prio);
+		param_get(param_find("CAL_MAG0_PRIO"), &mag0_prio);
+		param_get(param_find("CAL_MAG1_PRIO"), &mag1_prio);
+		param_get(param_find("CAL_MAG2_PRIO"), &mag2_prio);
+		param_get(param_find("CAL_MAG3_PRIO"), &mag3_prio);
+
+		const int32_t mag_param_priorities[MAX_NUM_MAGS] {mag0_prio, mag1_prio, mag2_prio, mag3_prio};
+		const int32_t imu_param_priorities[MAX_NUM_IMUS] {acc0_prio, acc1_prio, acc2_prio, acc3_prio};
+		// Define two arrays to store the sensor priorities and the original sensor instance number
+		int32_t mag_priorities[MAX_NUM_MAGS] {};
+		uint8_t mag_priorities_sensor_indexes[MAX_NUM_MAGS] {};
+		int32_t imu_priorities[MAX_NUM_IMUS] {};
+		uint8_t imu_priorities_sensor_indexes[MAX_NUM_IMUS] {};
+
+
+		uint8_t valid_mag_instances = 0;
+		uint8_t valid_imu_instances = 0;
+
+		// Initialize MAGs pair array
+		for (int i = 0; i < MAX_NUM_MAGS; i++) {
+			mag_priorities[i] = mag_param_priorities[i];
+			mag_priorities_sensor_indexes[i] = i;
+
+			// Priority higher than Disabled
+			if (mag_param_priorities[i] > 0) {
+				valid_mag_instances++;
+			}
+
+		}
+
+		// Initialize IMUs pair array
+		for (int i = 0; i < MAX_NUM_MAGS; i++) {
+			imu_priorities[i] = imu_param_priorities[i];
+			imu_priorities_sensor_indexes[i] = i;
+
+			// Priority higher than Disabled
+			if (imu_param_priorities[i] > 0) {
+				valid_imu_instances++;
+			}
+
+		}
+
+		// Inverse sort by sensor priority while keeping track of the original sensor number
+		sortPair(mag_priorities, mag_priorities_sensor_indexes, MAX_NUM_MAGS, true);
+		sortPair(imu_priorities, imu_priorities_sensor_indexes, MAX_NUM_IMUS, true);
+
+		for (int i = 0; i < MAX_NUM_MAGS; i++) {
+			PX4_INFO("MAGs priorities - i: %d  MAG%u  Priority: %" PRId32, i, mag_priorities_sensor_indexes[i],
+				 mag_priorities[i]);
+		}
+
+		PX4_INFO("Initialized IMUs (positive priority): %d", valid_imu_instances);
+
+		for (int i = 0; i < MAX_NUM_IMUS; i++) {
+			PX4_INFO("IMUs priorities - i: %d  ACC%u  Priority: %" PRId32, i, imu_priorities_sensor_indexes[i],
+				 imu_priorities[i]);
+		}
+
 		const hrt_abstime time_started = hrt_absolute_time();
-		const int multi_instances = math::min(imu_instances * mag_instances, static_cast<int32_t>(EKF2_MAX_INSTANCES));
-		int multi_instances_allocated = 0;
+		// Number of instances is limited by the number of valid sensors (not uninitialized or disabled) that are beeing published
+		const uint8_t multi_instances = static_cast<uint8_t>(math::min(valid_imu_instances * mag_instances,
+						valid_imu_instances * static_cast<int32_t>(valid_mag_instances), static_cast<int32_t>(EKF2_MAX_INSTANCES)));
```

Comment:
You also want to be limited by the param directly, right? We should then take the min of the param and the valid instances for both imu and mags.
```
		multi_instances = min(min(imu_instances, valid_imu_instances) * min(mag_instances, valid_mag_instances), EKF2_MAX_INSTANCES);
```

Reply:
That's correct, I will modify this